Add http header feature to httpPost and httpGet methods

PasinduYeshan opened the pull request:
## Purpose

This pull request aims to resolve issue [[#15679](https://github.com/wso2/product-is/issues/15679)](https://github.com/wso2/product-is/issues/15679) - Enable sending headers with HTTP requests.

The problem identified was that the `httpPost` and `httpGet` methods in the Adaptive Authentication script lacked the ability to send HTTP headers along with the request. Furthermore, the `httpPost` method only supported JSON payloads.

## Goals

The goal of this pull request is to enhance the `httpPost` and `httpGet` methods to:

1. Allow sending HTTP headers with the request.
2. Support multiple content types for the `httpPost` payload, including `application/json`, and `application/x-www-form-urlencoded.`

## Approach

The solution introduces `headers`, `payloadData`, and `eventHandlers` as optional parameters to the `httpPost` and `httpGet` functions. Depending on the length of the `params` array, the functions extract `headers`, `payloadData`, and `eventHandlers` from it.

Additionally, the `httpPost` function has been extended to handle different payload types based on the `Content-Type` header.

## **Documentation**

### **`httpGet` Method**

The **`httpGet`** method sends an HTTP GET request to the provided endpoint URL. This method can handle headers and event handlers as optional parameters.

Parameters for **`httpGet`**:

1. **`epUrl`** (required): This is the endpoint URL where the HTTP GET request will be sent.
2. **`headers`** (optional): This is a map containing the request headers.
3. **`eventHandlers`** (optional): This is a map containing the event handlers. You can specify actions for **`onSuccess`** and **`onFail`** events.

### **`httpPost` Method**

The **`httpPost`** method sends an HTTP POST request to the provided endpoint URL. This method can handle payload data, headers, and event handlers as optional parameters.

Parameters for **`httpPost`**:

1. **`epUrl`** (required): This is the endpoint URL where the HTTP POST request will be sent.
2. **`payloadData`** (optional): This is a map containing the data you want to send with the HTTP POST request.
3. **`headers`** (optional): This is a map containing the request headers.
4. **`eventHandlers`** (optional): This is a map containing the event handlers. You can specify actions for **`onSuccess`** and **`onFail`** events.

## **Examples**

### **`httpGet` Example**

Here's an example of how you can use the **`httpGet`** method to send an HTTP GET request with headers and event handlers:

```jsx

var onLoginRequest = function(context) {
    // Define the endpoint URL
    var getEndpointUrl = 'https://webhook.site/4c09b592-a272-4a55-865f-7edf5a69326f';

    // Define the headers
    var headers = {
        "Accept": "application/json",
        "Authorization": "Bearer your-token"
    };

    // Define eventHandlers if needed
    var eventHandlers = {
        onSuccess: function (context, data) {
            Log.info('HTTP GET request successful');
        },
        onFail: function (context, data) {
            Log.error('HTTP GET request failed');
        }
    };

    // Call the httpGet function
    httpGet(getEndpointUrl, headers, eventHandlers);

    // Execute the next step in the authentication script
    executeStep(1);
};

```

### **`httpPost` Example**

Here's an example of how you can use the **`httpPost`** method to send an HTTP POST request with payload data, headers, and event handlers:

```jsx

var onLoginRequest = function(context) {
    // Define the endpoint URL
    var postEndpointUrl = 'https://webhook.site/4c09b592-a272-4a55-865f-7edf5a69326f';

    // Define the payload data
    var payloadData = {
        "username": "Marcel",
        "password": "supersecret"
    };

    // Define the headers
    var headers = {
        "Content-Type": "application/json"
    };

    // Define eventHandlers if needed
    var eventHandlers = {
        onSuccess: function (context, data) {
            Log.info('HTTP POST request successful');
        },
        onFail: function (context, data) {
            Log.error('HTTP POST request failed');
        }
    };

    // Call the httpPost function
    httpPost(postEndpointUrl, payloadData, headers, eventHandlers);

    // Execute the next step in the authentication script
    executeStep(1);
};

```

Kanapriya commented on components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/AbstractHTTPFunction.java:
```diff
@@ -90,9 +94,18 @@ protected void executeHttpMethod(HttpUriRequest request, Map<String, Object> eve
+                    if (response.getEntity() != null) {
+                        Header contentType = response.getEntity().getContentType();
+                        String jsonString = EntityUtils.toString(response.getEntity());
+                        if (contentType != null && contentType.getValue().contains("text/plain")) {
```
Let's refer "text/plain" from a constant file.

Kanapriya commented on components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPGetFunctionImpl.java:
```diff
@@ -39,10 +41,38 @@ public HTTPGetFunctionImpl() {
 
     @Override
-    public void httpGet(String epUrl, Map<String, Object> eventHandlers) {
+    public void httpGet(String epUrl, Object... params) {
```
```suggestion
    public void httpGet(String endpointURL, Object... params) {
```

PasinduYeshan replied:
Changed

Kanapriya commented on components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPGetFunctionImpl.java:
```diff
@@ -39,10 +41,38 @@ public HTTPGetFunctionImpl() {
+
+        switch (params.length) {
+            case 1:
+                if (params[0] instanceof Map) {
```
Shall we add more debug logs to each switch case ?, Having a debug logs will give valuable insight about the logic and this will help to troubleshooting the issues as well.

Kanapriya commented on components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPGetFunctionImpl.java:
```diff
@@ -39,10 +41,38 @@ public HTTPGetFunctionImpl() {
+                    headers = (Map<String, String>) params[0];
+                    eventHandlers = (Map<String, Object>) params[1];
+                } else {
+                    throw new IllegalArgumentException("Invalid argument type.");
```
Let's elaborate more on the error messages and lets have different meaningful error messages per defined cases.

PasinduYeshan replied:
Added in [f529f98](https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/140/commits/f529f9879dc8d5b7ac378837a5f1c06d1a1c4a2a).

The expression `entry.getValue().getClass().getName()` returns the class name 
jdk.nashorn.api.scripting.ScriptObjectMirror in the context of JavaScript Objects, which is not the expected result. Therefore, the error message has been refined to only mention the expected type:

`throw new IllegalArgumentException("Invalid argument types. Expected payloadData and eventHandlers " +
                            "(both of type Map<String, Object>) respectively.");`
                            
Further, I checked with the current implementation of other function and following is how the `executeStep` method logs the error. 

```
public void executeStep(int stepId, Object... params) {
       // blaa blaa
        if (params.length > 0) {
            // if there are any params provided, last one is assumed to be the event listeners
            if (params[params.length - 1] instanceof Map) {
                attachEventListeners((Map<String, Object>) params[params.length - 1]);
            } else {
                log.error("Invalid argument and hence ignored. Last argument should be a Map of event listeners.");
            }
        }
        if (params.length == 2) {
           
        }
    }
    ```

PasinduYeshan commented on components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPGetFunctionImpl.java:
```diff
@@ -39,10 +41,38 @@ public HTTPGetFunctionImpl() {
+        Map<String, Object> eventHandlers;
+        Map<String, String> headers = new HashMap<>();
+
+        switch (params.length) {
```
Check if params are not passed what happen

PasinduYeshan replied:
When the params are not passed, default case occurs, which throws an `IllegalArgumentException`.

PasinduYeshan commented on components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPGetFunctionImpl.java:
```diff
@@ -39,10 +41,38 @@ public HTTPGetFunctionImpl() {
+
+        headers.putIfAbsent(ACCEPT, TYPE_APPLICATION_JSON);
+        headers.entrySet().stream()
+                .filter(entry -> !StringUtils.isBlank(entry.getKey()) && !entry.getKey().equals("null"))
```
use isNotBalnk()

PasinduYeshan replied:
`isBlank` is changed to `isNotBlank`. 
There's no `notEquals` method.

PasinduYeshan commented on components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPPostFunctionImpl.java:
```diff
@@ -44,17 +52,68 @@ public HTTPPostFunctionImpl() {
+
+
+        if (MapUtils.isNotEmpty(payloadData)) {
+            //For the header "Content-Type : application/x-www-form-urlencoded" request body data is set to
```
Change this to multi line comment

PasinduYeshan commented:
Header validation is added and moved to the `AbstractHTTPFunction` class as it is being used in both `httpGetFunctionImpl` and `httpPostFunctionImpl`.
Further new `setHeaders` method is introduced inside the  `AbstractHTTPFunction` class and removed redundant code in `httpGetFunctionImpl` and `httpPostFunctionImpl`.
@DMHP

Kanapriya commented on components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/http/http-get-test-headers.xml:
```diff
@@ -0,0 +1,78 @@
+<!--
+  ~ Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org).
```
Change the licence header and the year

Kanapriya commented on components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/http/http-post-test-headers.xml:
```diff
@@ -0,0 +1,81 @@
+<!--
+  ~ Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org).
```
Chnage the licence header and the year

DMHP commented on components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/AbstractHTTPFunction.java:
```diff
@@ -90,21 +95,30 @@ protected void executeHttpMethod(HttpUriRequest request, Map<String, Object> eve
+                        if (contentType != null && contentType.getValue().contains(TYPE_TEXT_PLAIN)) {
+                            // For 'text/plain', put the response body into the JSON object as a single field.
+                            json = new JSONObject();
+                            json.put("response", jsonString);
```
Extract "response" to a constant.

PasinduYeshan replied:
Refactored in [053e760](https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/140/commits/053e760da0fb8c3528bc8ee40d08eb2b92243320)

DMHP commented on components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/AbstractHTTPFunction.java:
```diff
@@ -172,4 +186,32 @@ private String getParentDomainFromUrl(URI url) {
+
+    /**
+     * Validate the headers and return a Map<String, String> of headers.
+     * @param headers Map of headers.
```
Modify the java doc comment by adding the return statement.

PasinduYeshan commented on components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/AbstractHTTPFunction.java:
```diff
@@ -172,4 +186,32 @@ private String getParentDomainFromUrl(URI url) {
     }
+
+    /**
+     * Validate the headers and return a Map<String, String> of headers.
```
Added in [053e760](https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/140/commits/053e760da0fb8c3528bc8ee40d08eb2b92243320)

Kanapriya commented:
@PasinduYeshan : Are we align this feature with the unification process?

PasinduYeshan commented:
I tested this with `Asgardeo IS - v0.12.3098` pack.

ThaminduDilshan commented on components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPPostFunctionImpl.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ * Copyright (c) 2018, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```
```suggestion
 * Copyright (c) 2018, WSO2 LLC. (https://www.wso2.com).
```
Check other places

PasinduYeshan replied:
Addressed in https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/171.

PasinduYeshan commented on components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPGetFunctionImpl.java:
```diff
@@ -39,10 +41,37 @@ public HTTPGetFunctionImpl() {
+                }
+                break;
+            case 2:
+                if (params[0] instanceof Map && params[1] instanceof Map ) {
```
Addressed in https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/171.
